whois-update/src/main/java/net/ripe/db/whois/update/generator/SourceGenerator.java

```diff
@@ -21,30 +28,73 @@ public class SourceGenerator extends AttributeGenerator {
 
     private final AuthoritativeResourceData authoritativeResourceData;
 
+    private final RpslObjectDao rpslObjectDao;
     private final CIString source;
     private final CIString nonAuthSource;
 
     @Autowired
     public SourceGenerator(final AuthoritativeResourceData authoritativeResourceData,
                            @Value("${whois.source}") final String source,
-                           @Value("${whois.nonauth.source}") final String nonAuthSource) {
+                           @Value("${whois.nonauth.source}") final String nonAuthSource,
+                           final RpslObjectDao rpslObjectDao) {
         this.authoritativeResourceData = authoritativeResourceData;
         this.source = ciString(source);
         this.nonAuthSource = ciString(nonAuthSource);
+        this.rpslObjectDao = rpslObjectDao;
     }
 
     @Override
     public RpslObject generateAttributes(RpslObject originalObject, RpslObject updatedObject, Update update, UpdateContext updateContext) {
-        switch (updatedObject.getType()) {
+        switch (updatedObject.getType()){
+            case AS_SET:
+                return generateAsSetSource(updatedObject, update, updateContext);
             case AUT_NUM:
             case ROUTE:
             case ROUTE6:
                 return generateSource(updatedObject, update, updateContext);
             default:
                 return updatedObject;
+
         }
     }
 
+    private RpslObject generateAsSetSource(final RpslObject updatedObject, final Update update, final UpdateContext updateContext) {
+        if (update.getOperation() == Operation.DELETE) {
+            return updatedObject;
+        }
+        final String asSetKey = updatedObject.getKey().toString();
+
+        final boolean flatAsSet = !asSetKey.contains(":");
+        final CIString asSetSource = updatedObject.getValueForAttribute(AttributeType.SOURCE);
+
+        if (flatAsSet){
+            final RpslObject asSetObject = updateContext.getPreparedUpdate(update).getReferenceObject();
```

Comment:
Use `originalObject` instead, it's passed in the parent method

Reply:
Yes attributeGenrator does have original Object , sorry I suggested to get it from updateContext instead of database. I missed this

Reply:
ohh totally blind! thank you!